api/src/migrations/convert-translation-messages.js

```diff
@@ -0,0 +1,82 @@
+  { promisify } = require('util'),
+  db = require('../db-nano'),
+  properties = require('properties');
+
+module.exports = {
+  name: 'convert-translation-messages',
+  created: new Date(2018, 11, 8),
+  run: promisify((callback) => {
+	  db.request({
+      db: db.settings.db,
+      method: 'GET',
+      path: '_design/medic',
+    },
+    (err, result) => {
+      if (err) {
+        return callback(err);
+      }
+
+      let translationAttachmentKeys = Object.keys(result._attachments).filter(k => k.includes('translations'));
+      translationAttachmentKeys.forEach((translationAttachmentKey) => {
+        let translationMessageKey = translationAttachmentKey.match(/translations\/(.+).properties/)[1];
+        let translationMessageKeys = [translationMessageKey, translationMessageKey + '-backup'];
+        translationMessageKeys.forEach((translationMessageKey) => {
+          db.request({
+            db: db.settings.db,
+            method: 'GET',
+            path: translationMessageKey,
+          },
+          (err, translationMessageValues) => {
+            if (_.has(translationMessageValues, 'values')) {
+              db.request({
+                db: db.settings.db,
+                method: 'GET',
+                path: '_design/medic/' + translationAttachmentKey,
+              },
+              (err, translationAttachmentValue) => {
+                if (err) {
+                  return callback(err);
+                }
+
+                properties.parse(translationAttachmentValue.toString('utf8'), (err, translationAttachmentValues) => {
+                  if (err) {
+                    return callback(err);
+                  }
+
+                  let defaultTranslations = {};
+                  let customTranslations = {};
+                  Object.keys(translationMessageValues.values).forEach((translationMessageKey) => {
+                    if (_.has(translationAttachmentValues, translationMessageKey)) {
+                      defaultTranslations[translationMessageKey] = translationMessageValues.values[translationMessageKey];
+                    } else {
+                      customTranslations[translationMessageKey] = translationMessageValues.values[translationMessageKey];
+                    }
+                  });
+
+                  delete translationMessageValues.values;
+                  translationMessageValues.custom = customTranslations;
+                  translationMessageValues.default = defaultTranslations;
+
+                  db.request({
+                    db: db.settings.db,
+                    method: 'PUT',
+                    path: translationMessageKey,
+                    body: translationMessageValues
+                  },
+                  (err) => {
+                    if (err) {
+                      return callback(err);
+                    }
+                  });
+                });
+              });
+            }
+          });
+        });
+      });
+
+      callback();
```

Comment:
There are quite a few bugs with the async code flow here...

1. This callback is executed _before_ the migration is complete because it's not waiting for the async code to complete. This will cause a race condition in the app, but will also mean API starts up even if this migration fails (it should not).
2. The `return callback` statements in the `forEach` loops don't return from the migration function - they return from the function given to the `forEach` call. This means the callback may be invoked multiple times which isn't expected by the API startup code.

I recommend when you change this code to use PouchDB you use promises instead of callbacks. One of the reasons I really dislike callbacks is it's really easy to make these sorts of errors (I do it all the time). Once you're using Promises you can do things like `return Promise.all(myAsyncFunctions);` to make sure the Promise doesn't resolve until all the requests have completed.

Reply:
I have opt to use `async/await` which produces a more readable code (and less buggy)

Reply:
Anything is better than callbacks! `async/await` is perfect.

Reply:
I think this callback is no longer necessary - because it awaits on the `Promise.all` you can just return when the function returns. It's really confusing to me to have async/await, promises, and callbacks in one block. Definitely find a way to get rid of the callback.